openshift/default-catalog-consistency/pkg/check/catalog.go

```diff
@@ -0,0 +1,122 @@
+			var errs []error
+			for _, bundle := range cfg.Bundles {
+				var found bool
+				for _, prop := range bundle.Properties {
+					if prop.Type == "olm.bundle.object" {
+						found = true
+					}
+				}
+				if found {
+					errs = append(errs, fmt.Errorf("bundle %q in package %q has olm.bundle.object set",
+						bundle.Name, bundle.Package))
+				}
+			}
+
+			if len(errs) == 0 {
+				return nil
+			}
+
+			return errors.Join(errs...)
+		},
+	}
+}
+
+// CatalogHasValidMetadataForChannelHeads checks if the channel heads have olm.csv.metadata
+// OLM Package Server (used by tooling such as the kubectl operator plugin and the OpenShift console)
+// does require these properties to be able to serve metadata about the packages in a catalog.
+//
+// In order to satisfy the needs of the package server, catalog maintainers should use this property to
+// include the CSV metadata for all bundles that are channel heads.
+// As of `opm` version 1.28.0, this property is automatically generated by when migrating SQLite
+// catalogs with `opm migrate` and when rendering SQLite catalogs and registry+v1 bundles with `opm render`.
+// Catalogs containing `olm.csv.metadata` properties must be served by `opm` binaries
+// with version at least `1.28.0`.
+func CatalogHasValidMetadataForChannelHeads() CatalogCheck {
+	return CatalogCheck{
+		Name: "CatalogHasValidMetadataForChannelHeads",
+		Fn: func(ctx context.Context, cfg declcfg.DeclarativeConfig) error {
+			var errs []error
+			packages, err := declcfg.ConvertToModel(cfg)
+			if err != nil {
+				return fmt.Errorf("unable to covert: %w", err)
+			}
+
+			for _, pkg := range packages {
+				for _, ch := range pkg.Channels {
+					channelHead, err := ch.Head()
+					if err != nil {
+						errs = append(errs, fmt.Errorf(
+							"channel %q in package %q could not determine channel head: %v",
+							ch.Name, pkg.Name, err))
+						continue
+					}
+
+					var found bool
+					for _, prop := range channelHead.Properties {
+						if prop.Type == "olm.csv.metadata" {
+							found = true
+						}
+					}
+					if !found {
+						errs = append(errs,
+							fmt.Errorf("bundle %q in package %q is missing olm.csv.metadata",
+								channelHead.Name,
+								channelHead.Package.Name))
+
+						// If we have a channel head without olm.csv.metadata we need to verify if the olm.bundle.object
+						// property is set. If not, then it means that the opm render | migrate command could not
+						// do the migration.
+						found = false
+						for _, prop := range channelHead.Properties {
+							if prop.Type == "olm.bundle.object" {
+								found = true
+							}
+						}
+
+						if !found {
+							errs = append(errs,
+								fmt.Errorf("bundle %q in package %q is a head of channel and does not either "+
+									"olm.csv.metadata or olm.bundle.object set",
+									channelHead.Name,
+									channelHead.Package.Name))
+						}
+					}
+				}
+			}
+
```

Comment:
I don't think it matters but in the previous method you have:

```
if len(errs) == 0 {
   return nil
}
```

Reply:
we do not need it.
removed from the previous one 👍